runtime/src/storage/iouring.rs

```diff
@@ -21,17 +26,131 @@ impl Config {
+        }
+
+        // Wrap the operation with this unique token to identify it later
+        let op = op.user_data(token);
+
+        // Submit the operation to the ring
+        {
+            let mut ring = self.ring.lock().unwrap();
+            unsafe {
+                ring.submission()
+                    .push(&op)
+                    .map_err(|_| Error::WriteFailed)?;
+            }
+            ring.submit().map_err(|_| Error::WriteFailed)?;
+        }
+
+        // Notify the completion handler
+        self.work_available.notify_one();
+
+        // Wait for completion
+        let res = receiver.await.map_err(|_| Error::RecvFailed)?; // TODO danlaine: update error
+        Ok(res)
+    }
+
+    // Background task that polls for completions
+    pub async fn completion_handler(self: Arc<Self>) {
+        loop {
+            // Process any completions that might be ready
+            let found_finished_work = self.process_completions();
+
+            // Check if we still have pending operations
+            let has_unfinished_work = {
+                let operations = self.work.lock().unwrap();
+                !operations.is_empty()
+            };
+
+            if !has_unfinished_work && !found_finished_work {
+                // No pending operations and no completions found
+                // Wait for notification about new work
+                self.work_available.notified().await;
+            } else if !found_finished_work {
+                // We have pending operations but no completions were ready
+                // Yield briefly then check again
+                tokio::task::yield_now().await;
+            }
+            // If we found completions, loop immediately to check for more
+        }
+    }
+
+    fn process_completions(&self) -> bool {
+        let mut completed = Vec::new();
+        {
+            let mut ring = self.ring.lock().unwrap();
+            ring.submit().unwrap_or(0);
+
+            while let Some(cqe) = ring.completion().next() {
+                let token = cqe.user_data();
+                completed.push((token, cqe.result()));
+            }
+        }
+
+        if !completed.is_empty() {
+            let mut operations = self.work.lock().unwrap();
+            for (token, result) in completed {
+                if let Some(sender) = operations.remove(&token) {
+                    let _ = sender.send(result);
+                }
+            }
+            true
+        } else {
+            false
+        }
+    }
+}
+
 #[derive(Clone)]
 pub struct Storage {
-    lock: Arc<Mutex<()>>,
-    pub storage_directory: PathBuf,
+    storage_directory: PathBuf,
+    io_ring: Arc<IoUringRuntimeAdapter>,
 }
 
 impl Storage {
     pub fn new(config: Config) -> Self {
+        let adapter = Arc::new(IoUringRuntimeAdapter::new());
+
+        // Start the background task for handling completions
+        let adapter_clone = adapter.clone();
+        tokio::spawn(async move {
```

Comment:
In this crate, we should never use `tokio` (or another runtime) directly other than for implementing `spawn*`